src/Halogen/Components/Select/Utils/NonEmpty.purs

```diff
@@ -0,0 +1,40 @@
+module Halogen.Components.Select.Utils.NonEmpty where
+
+import Prelude
+
+import Control.Alternative (Alternative)
+import Data.NonEmpty (NonEmpty(), (:|), oneOf)
+import Data.Maybe as M
+import Data.Maybe.Unsafe as Mu
+import Data.Array as A
+
+-- uncons empty = Nothing
+-- uncons (pure x <|> tail) = Just { head: x, tail: tail }
+class (Alternative f) <= Uncons f where
+  uncons :: forall a. f a -> M.Maybe {head :: a, tail :: f a}
+
+instance arrayUncons :: Uncons Array where
+  uncons = A.uncons
+
+liftNonEmpty
+  :: forall f a b
+   . (Uncons f)
+  => (f a -> f b)
+  -> NonEmpty f a
+  -> NonEmpty f b
+liftNonEmpty fab nefa =
+  let
+    fa :: f a
+    fa = oneOf nefa
+
+    fb :: f b
+    fb = fab fa
+
+    mbUnconsed :: M.Maybe {head :: b, tail :: f b}
+    mbUnconsed = uncons fb
+
+    -- Safe, `isJust (uncons (pure x <|> tail)) <-- oneOf === (pure x <|> tail)`
+    unconsed :: {head :: b, tail :: f b}
+    unconsed = Mu.fromJust mbUnconsed
```

Comment:
It took me ages to figure that one out. I can totally see why you wanted to do it this way, it'd be great to be able to keep the non-empty invariant while somehow manipulating the inner structure, but I can't really think of a safe way of making it happen unfortunately :confused: